### PR TITLE
🎨 Docs Design and Structure changes

### PR DESCRIPTION
**Pull request by MaedahBatool**
Our documentation site requires a significant design and structural overhaul. While I am diligently working on the progress toward achieving that goal, I am currently implementing some preliminary design and structural changes. These changes are aimed at rendering our documentation visually appealing, contemporary, clean, and well-organized in the interim period.

This PR ships the following design and structural changes to our docs website:

### Docs landing:
image
image

### Docs Cody landing page with improved sidebar design to show sub-level docs
image

### Improved small screens' whitespace and removed TOC on small screens, which isn't required.
image

### Improved sidebar highlighting and sub-level design.
image

### Modern search bar UI
image

### Experimental menu icons, more to come
image

### `h2` level headings come with a border-top to separate segments; no need to use <hr /> elements
image

### TOC redesign, white-space, `On this page` semantics, and clear `Edit this page on GitHub` CTA with SVG Icon
image

### Improved breadcrumbs design
image

### Dark mode link color
image

### Light mode link color
image

## Test plan

* Check links

### CODE REVIEW

File: doc/index.md [hunk 1]
```diff
@@ -3,9 +3,9 @@
 Sourcegraph makes it easy to read, write, and fix code—even in big, complex codebases.
 
 - **Code search:** Search all of your repositories across all branches and all code hosts.
-- **Code intelligence:** Navigate code, find references, see code owners, trace history, and more.
-- **Fix and refactor:** Roll out large-scale changes to many repositories at once and track big migrations.
-- **AI code assistant:** Use Cody to read, write, and understand code faster.
+- **Code intelligence:** Navigate code, find references, see code owners, trace history, & more.
+- **Fix & refactor:** Roll out & track large-scale changes & migrations across repos at once.
+- **AI code assistant:** Use Cody to read, write, and understand entire codebas faster.
```
Comment (camdencheek) [inline doc/index.md:8]:
```suggestion
- **AI code assistant:** Use Cody to read, write, and understand entire codebase faster.
```

File: doc/sidebar.md [hunk 2]
```diff
@@ -10,9 +10,10 @@ Keep it as a single list with at most 2 levels. (Anything else may not render co
 
 -->
 
-- [Home](index.md)
-- [Getting started](getting-started/index.md)
-- [Tutorials](tutorials/index.md)
+<!-- - [Home](index.md) -->
+
+<!-- - [Getting Started](getting-started/index.md) -->
+<!-- - [Tutorials](tutorials/index.md) -->
```
Comment (camdencheek) [inline doc/sidebar.md:16]: There is a bunch of stuff in this PR that is commented out. Should we just remove it all if it's unused?

File: doc/index.md (hunk 1 above)
Comment (peterguy) [reply doc/index.md:8]: pluralize `codebases`?

File: doc/sidebar.md (hunk 2 above)
Comment (camdencheek) [reply doc/sidebar.md:16]: for sure! not blocking at all, just wanted to point it out in case it was not intentional

**Review by peterguy — APPROVED** [review]
Design thinks "this is So clean!" and the overall code changes look great!

Follow up with PRs to update the content (Sourcegraph --> Sourcegraph Code Search, for example) and fix typos.

**Review by ryphil — APPROVED** [review]

**Review by toolmantim — APPROVED** [review]
Lovely! 😻

File: doc/sidebar.md (hunk 2 above)
Comment (MaedahBatool) [reply doc/sidebar.md:16]: This is intentional, please don't remove just yet. I plan to migrate the sidebar to a better structure so keeping a history commented for a while.

File: doc/index.md (hunk 1 above)
Comment (MaedahBatool) [reply doc/index.md:8]: Ah dang, thanks for pointing it out. #55342 has the fix. Let's merge. 

Note to self: Implement automated typos checking.

File: doc/index.md [hunk 3]
```diff
@@ -3,9 +3,9 @@
 Sourcegraph makes it easy to read, write, and fix code—even in big, complex codebases.
 
 - **Code search:** Search all of your repositories across all branches and all code hosts.
-- **Code intelligence:** Navigate code, find references, see code owners, trace history, and more.
-- **Fix and refactor:** Roll out large-scale changes to many repositories at once and track big migrations.
-- **AI code assistant:** Use Cody to read, write, and understand code faster.
+- **Code intelligence:** Navigate code, find references, see code owners, trace history, & more.
+- **Fix & refactor:** Roll out & track large-scale changes & migrations across repos at once.
```
Comment (courier-new) [inline doc/index.md:7]: Is this an intentional stylistic change we're going for, replacing the "and"s with "&"s? No opinions here, just curious. 🙂 Will we want to apply this to other parts of the docs eventually, too?

**Review by courier-new — COMMENTED** [review]
So this is looking awesome and I absolutely LOVE seeing the momentum on this already!!!

First of all, code changes look okay to me! 👍

I'm not sure I agree with the decision to universally hide the ToC on mobile. I admit I don't think any of us have the best understanding of how much/to what end users interact with our docs on mobile devices, but some of the longer reference-type docs pages _really_ rely on the ToC, like the [search query language page](https://docs.sourcegraph.com/code_search/reference/language) or [batch change YAML page](https://docs.sourcegraph.com/batch_changes/references/batch_spec_yaml_reference), and it seems disappointing that we lose that on even the narrow computer browser window size. I think the ideal for me would be a collapsible drawer that's default expanded on large screens and default collapsed on narrower or mobile. 

Lastly, in terms of the visual changes, obviously the docs site != the application, and I'm not actually sure how closely we want to adhere to the Wildcard design system for the docs site, but I just wanted to offer some feedback on the visuals and in particular about differences in the design language that I noticed with these changes. I don't mean to advocate for changing things _just_ to match Wildcard, haha, but I just thought this ought to be a part of the review process, so that we can be better aligned on design principles for the docs site going forward. 😄

- The search input field looks kinda disabled to me. This is partially because I'm conditioned that medium gray background with no border signifies disabled input, since that's how we do it from within the application itself (see [this Storybook story](https://storybook.sgdev.org/?path=/story/wildcard-input--simple) for an example), but I'm also pretty sure that's a more universally-recognized design pattern, too. This might be one case where I'd actually advocate for using a different style of input field that's more consistent with the application.
- The cyan links look 🔥 but typically we reserve cyan for "info" and just use the medium blue links in the rest of the application based on the [Wildcard semantic color guidelines](https://www.figma.com/file/NIsN34NH7lPu04olBzddTw/%E2%9C%B3%EF%B8%8F-Wildcard-Design-System?type=design&node-id=5601-64976&mode=design&viewport=-193,-2908,0.5&t=OxPwJMYYp6gSFEdZ-0). But idk also the cyan looks cool so 🤷‍♀️
  - Some places still appear to be using the paler blue on dark mode, though, such as the search query language reference page:
      image

  - as well as the getting started blocks on "Code Search", "Batch Changes", and "CLI", which I assume is unintentional:
      image
  - Side note, the spacing on the boxes and text there looks sorta clunky?
- That secondary dark purply gray color for links like these in dark mode looks kinda out of place next to the cyan:
  image
- Should the vertical bar + horizontal rules be the same color?
   image

Again super psyched about this, please feel free to tag me for frontend review of future PRs, too, if you'd like. 😄

**Comment by courier-new** [comment]
Ah I missed that you just merged this! 🙈 I meant to get you this feedback earlier today and it just kept getting pushed later. Apologies for the late feedback!! 🙇‍♀️

**Comment by unknwon** [comment]
to the PR and @courier-new's thorough review

**Pull request merged**